compiler/rustc_parse/src/parser/diagnostics.rs

```diff
@@ -1112,6 +1165,150 @@ impl<'a> Parser<'a> {
+        op_span: Span,
+    ) -> PResult<'a, P<Expr>> {
+        let mut err = self.struct_span_err(
+            op_span,
+            &format!("Rust has no {} {} operator", kind.fixity, kind.op.name()),
+        );
+        err.span_label(op_span, &format!("not a valid {} operator", kind.fixity));
+
+        let help_base_case = |mut err: DiagnosticBuilder<'_>, base| {
+            err.help(&format!("use `{}= 1` instead", kind.op.chr()));
+            err.emit();
+            Ok(base)
+        };
+
+        // (pre, post)
+        let spans = match kind.fixity {
+            UnaryFixity::Pre => (op_span, base.span.shrink_to_hi()),
+            UnaryFixity::Post => (base.span.shrink_to_lo(), op_span),
+        };
+
+        match kind.standalone {
+            Some(true) => self.inc_dec_standalone_recovery(&mut err, kind, spans, false),
+            Some(false) => {
+                let Ok(base_src) = self.span_to_snippet(base.span)
+                    else { return help_base_case(err, base) };
+                match kind.fixity {
+                    UnaryFixity::Pre => {
+                        self.prefix_inc_dec_suggest(base_src, &mut err, kind, spans)
+                    }
+                    UnaryFixity::Post => {
+                        self.postfix_inc_dec_suggest(base_src, &mut err, kind, spans)
+                    }
+                }
+            }
+            None => {
+                let Ok(base_src) = self.span_to_snippet(base.span)
+                    else { return help_base_case(err, base) };
+                match kind.fixity {
+                    UnaryFixity::Pre => {
+                        self.prefix_inc_dec_suggest(base_src, &mut err, kind, spans)
+                    }
+                    UnaryFixity::Post => {
+                        self.postfix_inc_dec_suggest(base_src, &mut err, kind, spans)
+                    }
+                }
+                self.inc_dec_standalone_recovery(&mut err, kind, spans, true)
+            }
+        }
+        Err(err)
+    }
+
+    fn prefix_inc_dec_suggest(
+        &mut self,
+        base_src: String,
+        err: &mut DiagnosticBuilder<'a>,
+        kind: IncDecRecovery,
+        (pre_span, post_span): (Span, Span),
+    ) {
+        err.multipart_suggestion(
+            &format!("use `{}= 1` instead", kind.op.chr()),
+            vec![
+                (pre_span, "{ ".to_string()),
+                (post_span, format!(" {}= 1; {} }}", kind.op.chr(), base_src)),
+            ],
+            Applicability::MachineApplicable,
+        );
+    }
+
+    fn postfix_inc_dec_suggest(
+        &mut self,
+        base_src: String,
+        err: &mut DiagnosticBuilder<'a>,
+        kind: IncDecRecovery,
+        (pre_span, post_span): (Span, Span),
+    ) {
+        let mut msg = format!("use `{}= 1` instead", kind.op.chr());
+        if base_src.trim() == "tmp" {
+            msg.push_str(" (rename `tmp` so it doesn't conflict with your variable)");
```

Comment:
Wording could probably be improved here.

Reply:
If we can detect this, why can't we just change the variable name in our suggestion?

Reply:
I thought about doing that but worried it might be overkill. My idea was to have a loop where we add underscores until the name doesn't match. Alternatively, it could append a number. Personally I prefer the underscore approach because I doubt it'll require more than one or two underscores, whereas `tmp4` could be a real variable.

Reply:
I just realized that there's no need for a loop because we're only comparing against one variable 🤦